src/mpid/ch4/netmod/ofi/ofi_am_impl.h

```diff
@@ -276,44 +335,115 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_am_isend_short(int rank, MPIR_Comm * comm
     MPIR_Assert(data_sz < (1ULL << MPIDI_OFI_AM_DATA_SZ_BITS));
     MPIR_Assert((uint64_t) comm->rank < (1ULL << MPIDI_OFI_AM_RANK_BITS));
 
-    msg_hdr = &MPIDI_OFI_AMREQUEST_HDR(sreq, msg_hdr);
+    msg_hdr = &send_req->req_hdr.msg_hdr;
     msg_hdr->handler_id = handler_id;
-    msg_hdr->am_hdr_sz = MPIDI_OFI_AMREQUEST_HDR(sreq, am_hdr_sz);
+    msg_hdr->am_hdr_sz = send_req->req_hdr.am_hdr_sz;
     msg_hdr->data_sz = data_sz;
     msg_hdr->am_type = MPIDI_AMTYPE_SHORT;
     msg_hdr->seqno = MPIDI_OFI_am_fetch_incr_send_seqno(comm, rank);
     msg_hdr->fi_src_addr
         = MPIDI_OFI_comm_to_phys(MPIR_Process.comm_world, MPIR_Process.comm_world->rank, 0, 0);
+    msg_hdr->sreq_ptr = sreq;
 
-    iov = MPIDI_OFI_AMREQUEST_HDR(sreq, iov);
+    iov = send_req->req_hdr.iov;
 
     iov[0].iov_base = msg_hdr;
     iov[0].iov_len = sizeof(*msg_hdr);
 
-    iov[1].iov_base = MPIDI_OFI_AMREQUEST_HDR(sreq, am_hdr);
-    iov[1].iov_len = MPIDI_OFI_AMREQUEST_HDR(sreq, am_hdr_sz);
+    iov[1].iov_base = send_req->req_hdr.am_hdr;
+    iov[1].iov_len = send_req->req_hdr.am_hdr_sz;
 
     iov[2].iov_base = (void *) data;
     iov[2].iov_len = data_sz;
 
     MPIR_cc_incr(sreq->cc_ptr, &c);
-    MPIDI_OFI_AMREQUEST(sreq, event_id) = MPIDI_OFI_EVENT_AM_SEND;
+    send_req->event_id = MPIDI_OFI_EVENT_AM_SEND;
     MPIDI_OFI_CALL_RETRY_AM(fi_sendv(MPIDI_OFI_global.ctx[0].tx, iov, NULL, 3,
                                      MPIDI_OFI_comm_to_phys(comm, rank, 0, 0),
-                                     &MPIDI_OFI_AMREQUEST(sreq, context)), sendv);
+                                     &send_req->context), sendv);
   fn_exit:
     MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_OFI_AM_ISEND_SHORT);
     return mpi_errno;
   fn_fail:
     goto fn_exit;
 }
 
-MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_deferred_am_isend_enqueue(int rank, MPIR_Comm * comm,
+MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_am_isend_pipeline(int rank, MPIR_Comm * comm, int handler_id,
+                                                         const void *data, MPI_Aint seg_sz,
+                                                         MPI_Aint data_sz, MPIR_Request * sreq,
+                                                         MPIDI_OFI_am_send_request_t * send_req)
+{
```

Comment:
This function sends the header + data, right? What is the difference from `send_short`?

Reply:
For first segment, it is am_hdr + data, the rest segments are data only.